calico-vpp-agent/policy/policy_server.go

```diff
@@ -317,87 +369,126 @@ func (s *Server) Stop() {
+				}
+			}
+			if err != nil {
+				s.log.WithError(err).Error("Error processing update from felix, restarting")
+				conn.Close()
+				s.felixRestarted <- true
+				// TODO: Restart VPP as well? State is left over there...
+				return
 			}
+		case tunnelChange := <-s.routingServer.ConnectivityServer.GetTunnelChangeChan():
 			var pending bool
 			if s.state == StateSyncing {
 				pending = true
 			} else if s.state == StateInSync {
 				pending = false
 			} else {
-				s.log.Errorf("Got message %+v but not in syncing or synced state", m)
-				conn.Close()
-				s.felixRestarted <- true
-				return
+				s.log.Errorf("Got tunnel change %+v but not in syncing or synced state", tunnelChange)
+				continue
 			}
-			switch m := msg.(type) {
-			case *proto.IPSetUpdate:
-				err = s.handleIpsetUpdate(m, pending)
-			case *proto.IPSetDeltaUpdate:
-				err = s.handleIpsetDeltaUpdate(m, pending)
-			case *proto.IPSetRemove:
-				err = s.handleIpsetRemove(m, pending)
-			case *proto.ActivePolicyUpdate:
-				err = s.handleActivePolicyUpdate(m, pending)
-			case *proto.ActivePolicyRemove:
-				err = s.handleActivePolicyRemove(m, pending)
-			case *proto.ActiveProfileUpdate:
-				err = s.handleActiveProfileUpdate(m, pending)
-			case *proto.ActiveProfileRemove:
-				err = s.handleActiveProfileRemove(m, pending)
-			case *proto.HostEndpointUpdate:
-				err = s.handleHostEndpointUpdate(m, pending)
-			case *proto.HostEndpointRemove:
-				err = s.handleHostEndpointRemove(m, pending)
-			case *proto.WorkloadEndpointUpdate:
-				err = s.handleWorkloadEndpointUpdate(m, pending)
-			case *proto.WorkloadEndpointRemove:
-				err = s.handleWorkloadEndpointRemove(m, pending)
-			case *proto.HostMetadataUpdate:
-				err = s.handleHostMetadataUpdate(m, pending)
-			case *proto.HostMetadataRemove:
-				err = s.handleHostMetadataRemove(m, pending)
-			case *proto.IPAMPoolUpdate:
-				err = s.handleIpamPoolUpdate(m, pending)
-			case *proto.IPAMPoolRemove:
-				err = s.handleIpamPoolRemove(m, pending)
-			case *proto.ServiceAccountUpdate:
-				err = s.handleServiceAccountUpdate(m, pending)
-			case *proto.ServiceAccountRemove:
-				err = s.handleServiceAccountRemove(m, pending)
-			case *proto.NamespaceUpdate:
-				err = s.handleNamespaceUpdate(m, pending)
-			case *proto.NamespaceRemove:
-				err = s.handleNamespaceRemove(m, pending)
-			default:
-				s.log.Warnf("Unhandled message from felix: %v", m)
+			state := s.currentState(pending)
+			for _, h := range state.HostEndpoints {
+				if tunnelChange.ChangeType == connectivity.AddChange {
+					newTunnel := true
+					for _, v := range h.TunnelSwIfIndexes {
+						if v == tunnelChange.Swifindex {
+							newTunnel = false
+						}
+					}
+					if newTunnel {
+						h.TunnelSwIfIndexes = append(h.TunnelSwIfIndexes, tunnelChange.Swifindex)
+						s.log.Infof("Configuring policies on added tunnel [%d]", tunnelChange.Swifindex)
+						err = s.vpp.ConfigurePolicies(tunnelChange.Swifindex, h.currentForwardConf)
```

Comment:
We should call ConfigurePolicies only if we aren't `pending`? Unless we have the guarantee that `h.currentForwardConf` always make this call do nothing when `h` hasn't been created (even if, for instance, it was created once and then VPP restarted).

Also, these few lines look like details that are internal to the HostEndpoint logic, and I think it would be better to put them in a method on the HostEndpoint struct.